File: src/contests/cg/puzzle/ocaml/max-area.ml
(* 🎮 CodinGame Puzzle - max-area *)
(* https://www.codingame.com/training/easy/max-area *)

let () =
  let _ = read_line () in
  let a =
    Array.of_list
      (List.map int_of_string
         (String.split_on_char ' ' (String.trim (read_line ()))))
  in
  let left = ref 0 and right = ref (Array.length a - 1) and best = ref 0 in
  while !left < !right do
    let h = min a.(!left) a.(!right) in
    let area = h * (!right - !left) in
    if area > !best then best := area;
    if a.(!left) < a.(!right) then incr left else decr right
  done;
  Printf.printf "%d\n" !best

File: src/contests/cg/puzzle/c/max-area.c
// 🎮 CodinGame Puzzle - max-area
// https://www.codingame.com/training/easy/max-area

#include <stdio.h>
#include <stdlib.h>

int main() {
    int n;
    scanf("%d", &n);
    long long *a = malloc(n * sizeof(long long));
    for (int i = 0; i < n; i++) scanf("%lld", &a[i]);
    int left = 0, right = n - 1;
    long long best = 0;
    while (left < right) {
        long long h = a[left] < a[right] ? a[left] : a[right];
        long long area = h * (right - left);
        if (area > best) best = area;
        if (a[left] < a[right]) left++;
        else right--;
    }
    printf("%lld\n", best);
    free(a);
    return 0;
}
